File: JavaScript/test/bfs_flood_fill.js
const chai = require('chai');
const expect = chai.expect;

const floodFill = (image, startRow, startCol, newColor) => {
  let startColor = image[startRow][startCol];
  let queue = [];
  queue.push([startRow, startCol]);
  let seen = new Set();

  const inBounds = ([r, c] = child) => r >= 0 && c >= 0 && r < image.length && c < image[0].length;

  while(queue.length){
    let current = queue.pop();
    if(seen.has(JSON.stringify(current))) continue;
    seen.add(JSON.stringify(current));

    let [row, col] = current;
    if(image[row][col] !== startColor) continue;
    image[row][col] = newColor;
    let children = [
      [row + 1, col],
      [row - 1, col],
      [row, col + 1],
      [row, col - 1]
    ].filter(c => inBounds(c));
    children.forEach(c => queue.push(c));
  }
  return image;
};

describe('Flood fill', () => {
  it('Basic BFS', () => {
    let image = [
      [1, 1, 1],
      [1, 1, 0],
      [1, 0, 1]
    ];

    let [row, col] = [1, 1];
    let newColor = 2;
    let expected = [
      [2, 2, 2],
      [2, 2, 0],
      [2, 0, 1]
    ];

    let actual = floodFill(image, row, col, newColor);
    expect(actual).to.deep.equal(expected);
  });
  it('Basic BFS 2', () => {
    return;
    let image = [
      [7, 7, 1],
      [7, 7, 0],
      [7, 0, 1]
    ];

    let [row, col] = [1, 1];
    let newColor = 2;
    let expected = [
      [2, 2, 1],
      [2, 2, 0],
      [2, 0, 1]
    ];

    let actual = floodFill(image, row, col, newColor);
    expect(actual).to.deep.equal(expected);
  });
});

File: Perl/t/bfs_flood_fill.pl
#!/usr/bin/env perl

use strict;
use warnings;
use version; our $VERSION = qv('1.0.0');

use Test::More::Behaviour;
use Data::Dumper;

sub flood_fill {
    my ( $image_ref, $start_row, $start_col, $new_color ) = @_;
    my @image       = @{$image_ref};
    my $start_color = $image[$start_row][$start_col];

    my @queue = ();
    push @queue, [ $start_row, $start_col ];

    my %seen = ();

    my $in_bounds = sub {
        my ( $r, $c ) = @{ shift; };

        my $result =
             $r >= 0
          && $c >= 0
          && $r < @{$image_ref}
          && $c < @{ @{$image_ref}[0] } ? 1 : 0;

        return $result;
    };

    while (@queue) {
        my $current = pop @queue;
        if ( exists $seen{ Dumper($current) } ) { next; }
        $seen{ Dumper($current) } = 1;
        my ( $row, $col ) = @{$current};
        if ( $image[$row][$col] != $start_color ) { next; }
        $image[$row][$col] = $new_color;
        my @children = (
            [ $row + 1, $col ],
            [ $row - 1, $col ],
            [ $row,     $col + 1 ],
            [ $row,     $col - 1 ],
        );
        @children = grep { $in_bounds->($_) } @children;

        foreach my $child (@children) {
            push @queue, $child;
        }
    }

    return @image;
}

describe 'Flood fill' => sub {
    it 'Basic flood fill' => sub {
        my @image = ( [ 1, 1, 1 ], [ 1, 1, 0 ], [ 1, 0, 1 ] );

        my ( $row, $col ) = ( 1, 1 );
        my $new_color = 2;
        my @expected  = ( [ 2, 2, 2 ], [ 2, 2, 0 ], [ 2, 0, 1 ] );
        my @got       = flood_fill( \@image, $row, $col, $new_color );
        is_deeply( \@got, \@expected );
    }

};

done_testing;
1;
